Automatic Formatting -> core

emilyyyylime opened the pull request:
This PR has a couple of commits that should be reviewed separately from the rest.
These are:
- **Emily: Change indentation to tabs everywhere**
- **Emily: Apply new formatting to all code**

Because they apply formatting changes to the entire codebase at once, they should be inspected for a format we are happy with, and otherwise ignored.

YoniKiriaty commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -64,16 +63,16 @@ public void simulationPeriodic() {
+		case REPLAY -> {
+			setUseTiming(false); // Run as fast as possible
+			LoggerUtils.startReplayLogger();
+		}
```
switch case need indent

YoniKiriaty replied:
Yoni (The actual Yoni) - facts

YoniKiriaty commented on src/main/java/frc/utils/calibration/swervecalibration/WheelRadiusCharacterization.java:
```diff
@@ -92,8 +96,7 @@ public void end(boolean interrupted) {
 			Logger.recordOutput(WheelRadiusConstants.LOG_PATH, "Not enough data for characterization");
-		}
-		else {
+		} else {
```
yoav - i think the else needs to be in a new line

YoniKiriaty replied:
Yoni - i dont care

YHerm reviewed (commented):
can you add that it will do 1 line space at start and end of classes?

YHerm commented on .scripts/pre-commit:
```diff
@@ -1,30 +1,24 @@
-    echo -e "${GREEN}Static analysis found no problems. Running add${NO_COLOR}"
-    git add .
+if [ "$status" = 0 ]; then
+    echo "Static analysis succeeded. Updating commit."
```
return the green

emilyyyylime replied:
My reasoning was that a successful run should not stand out when committing, so you can focus only on anything going wrong

YHerm replied:
I prefer it to be easy to see good and bad runs (honesty i really like colors)

bossyshrimp68 replied:
agree with yoav

YHerm commented on codestyle.xml:
can you add that i will do 1 line space at start and end of classes

YHerm commented on .scripts/pre-commit:
```diff
@@ -1,30 +1,24 @@
-RED='\033[0;31m'
-NO_COLOR='\033[0m'
+RED=$(printf '\033[0;31m')
+CLEAR=$(printf '\033[0m')
```
consider rename to no color or white or something like that

emilyyyylime replied:
Actually I see it referred to by 'reset' most often around the web. Are you cool with that?

YHerm replied:
ok

YHerm commented on .scripts/pre-commit:
```diff
@@ -1,30 +1,24 @@
-# Run spotless apply
-./gradlew spotlessApply
+# Find all files to be committed
+readarray -d '' changes -u <(git diff -z --staged --name-only)
```
check the changes var part again

YHerm replied:
i dont sure its working, i sent to you in whatsapp the output i got

YHerm commented on src/main/java/frc/utils/applicationsutils/CMDHandler.java:
```diff
@@ -5,53 +5,52 @@
+	public static void runCMDCommand(String command) {
+		Runtime runtime = Runtime.getRuntime();
+		try {
+			runtime.exec(new String[] { APPLICATION, CMD_DIRECTORY, command });
```
```suggestion
			runtime.exec(new String[] {APPLICATION, CMD_DIRECTORY, command});
```

maybe remove these

YoniKiriaty replied:
Facts

emilyyyylime commented:
Basically I'm mostly waiting for Romi and/or Raz/mentors to add their thoughts of the formatting changes (both initial and added after-the fact).

bossyshrimp68 commented on src/main/java/frc/utils/joysticks/SmartJoystick.java:
```diff
@@ -8,205 +8,207 @@
+	 * <p>
+	 * @formatter:off
+	 *			 Before:									After:
+	 *			  (1,0)
+	 *		   *****@*******	  @ (1,1)				*************
+	 *	  *****		   *****					*****	(1,0)	 *****
+	 *	 ***				   ***				***  -------@-------@   ***
+	 *	**					   **			  **   |		  (1,1)|	**
+	 *   **						 **			**	|			   |	 **
+	 *   **						 *@ (0,1)	  **	|		  (0,1)@	 **
+	 *   **						 **			**	|			   |	 **
+	 *   **						 **			**	|			   |	 **
+	 *	**					   **			  **   |_______________|	**
+	 *	 ***				   ***				***					***
+	 *	   *****		   *****					*****			*****
+	 *		   *************							*************
```
the drawing..

bossyshrimp68 reviewed (changes requested):
good job!!

emilyyyylime commented on .gitignore:
```diff
@@ -48,14 +48,17 @@
 *.tar.gz
 *.rar
 
+# Eclipse generated files
```
See below; the `.factorypath` file is added when formatting with eclipse-jdt. Actually it might also be reasonable to add [these lines](https://github.com/GreenBlitz/GB-Robot-Template/pull/189/files/9139b62b3a556a8a3719c752ca2cd20ca48085f1#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947R156-R161) under this header, as they are also generated by eclipse-jdt, though not when formatting (so if you use the Java VSCode extension they will be generated, as it uses eclipse-jdt).

bossyshrimp68 commented on build.gradle:
whats going on here?

emilyyyylime commented:
This is ready for merge :+1:

In the new workflow, every time someone tries to commit, spotless first checks if their code is formatted correctly. If not, the commit is cancelled with the following message:
![Spotless check failed. Cancelling commit... / Please fix code formatting before committing.](https://github.com/GreenBlitz/GB-Robot-Template/assets/40892795/876e3e55-e396-4e80-818b-c95ac41ce034).

(To update your local clone, you must at least once run `./gradlew installLocalGitHookPreCommit`—but this automatically happens when building).

emilyyyylime commented:
@bossyshrimp68

YHerm reviewed (changes requested):
What's going on with the behavior changing?

YHerm commented on .scripts/git-hooks/post-commit:
```diff
@@ -0,0 +1,41 @@
+pre_commit_head=$(cat .pre_commit_head)
+rm .pre_commit_head
+
+# Keep unstaged files somewhere safe
```
```suggestion
# Keep unstaged files somewhere safe, and pop them when script exit
```

emilyyyylime commented on .scripts/git-hooks/pre-commit:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+git rev-parse HEAD > .pre_commit_head
```
It's used in the post commit script both to figure out if it needs to run and to know where to return to if it fails

emilyyyylime commented on build.gradle:
```diff
@@ -148,28 +149,47 @@ tasks.register('checkAkitInstall', JavaExec) {
+        def linkDir = file("$rootDir/.git/hooks/").toPath()
+        def gitHooks = fileTree("$rootDir/.scripts/git-hooks/")
+        gitHooks.forEach { File target ->
+            def link = linkDir.resolve(target.toPath().getFileName())
```
google symbolic link

YHerm replied:
I meant consider renaming to better name, what the link represents?

emilyyyylime replied:
it's... a path to a symbolic link

YHerm replied:
Symbolic link to where?

emilyyyylime commented on build.gradle:
```diff
@@ -148,28 +149,47 @@ tasks.register('checkAkitInstall', JavaExec) {
+
+                Files.createSymbolicLink(link, target.toPath())
+
+                println("---Installed Symbolic Link ${rootDir.relativize(link)} -> ${rootDir.relativize(target)} ---")
```
idk what else I could add, I factored out the relativize calls in a later commit

emilyyyylime commented on .scripts/git-hooks/post-commit:
```diff
@@ -0,0 +1,41 @@
+
+set -e
+trap 'echo "${RED}post-commit script error on line ${LINENO}! Cannot continue.
+(If you have missing local changes, check the git stash).${RESET}"; git reset --soft "$pre_commit_head"' ERR
```
If you do `commit --amend` the pre-commit head will be the same commit as the current commit, there are probably other edge-cases that trigger this too.

YHerm replied:
Can the code exit with error after the ammend?

emilyyyylime replied:
it should never happen but I made sure the code is resilient to errors

YHerm replied:
Where does it create the file? Where in the computer

emilyyyylime replied:
Alongside the `.git` folder, so in the root of the project. Of course it gets removed before you could notice and I've added it to the .gitignore list so it shouldn't ever be noticeable

emilyyyylime commented on gradle/wrapper/gradle-wrapper.jar:
I updated this using gradle and it all seems to work

YHerm commented on build.gradle:
```diff
@@ -148,26 +149,55 @@ tasks.register('checkAkitInstall', JavaExec) {
+        def gitHooks = fileTree("$rootDir/.scripts/git-hooks/")
+        gitHooks.forEach { File target ->
+            def link = linkDir.resolve(target.toPath().getFileName())
+
+            if (!Files.isSymbolicLink(link)) {
+                if (Files.exists(link)) {
+                    Files.delete(link)
+                }
+
+                def rel = { f -> rootDir.toPath().relativize(f) }
+                Files.createSymbolicLink(link, target.toPath())
+
+
+                println("---Installed Symbolic Link ${rel(link)} -> ${rel(target.toPath())} ---")
+            }
+        }
```
so much line spaces, also give rel an informative name

YHerm replied:
the name link and relativePath doesn't tell a lot to someone who didn't write this code. Try to name it with more informative names

emilyyyylime replied:
I've factored this out to a separate `ensureSymlink` function, I think the names are more fitting given that it's built for a more generic case

YHerm commented on build.gradle:
```diff
@@ -148,26 +149,55 @@ tasks.register('checkAkitInstall', JavaExec) {
+    outputs.file('.git/config')
+}
+
+tasks
+        .matching { it.group != gitConfig }
+        .configureEach {
+            it.dependsOn tasks.matching {
+                it.group == gitConfig
+            }
+        }
```
```suggestion
tasks .matching { it.group != gitConfig } .configureEach {
            it.dependsOn tasks.matching {
                it.group == gitConfig
            }
        }
```

YHerm commented on build.gradle:
```diff
@@ -148,26 +149,55 @@ tasks.register('checkAkitInstall', JavaExec) {
+            }
+        }
+
+if (System.getenv().containsKey('CI') || !file('.git').isDirectory()) {
+    tasks
+            .matching { it.group == gitConfig }
+            .configureEach { it.enabled = false }
```
maybe make it one line, also when does it run?

emilyyyylime replied:
When running in CI or where there is no git directory

YHerm replied:
i meant not the inside when it is actually calling the if?

YHerm commented on build.gradle:
```diff
@@ -201,22 +231,20 @@ tasks.register('SetConnectionTypeIP') {
+        eclipse().configFile("codestyle.xml")
+        formatAnnotations()
     }
     groovyGradle {
```
maybe remove gradle mis and json formats

emilyyyylime replied:
It seems like a good idea to me to leave them, I removed the forced indent for misc tho

YHerm replied:
so you don't prefer the indent with tabs for them?

YHerm commented on build.gradle:
```diff
@@ -148,33 +149,60 @@ tasks.register('checkAkitInstall', JavaExec) {
-    }
-    doLast {
-        println("---Added Post Checkout Git Hook Script---")
+def ensureSymlink(link, target) {
```
```suggestion
def ensureSymbolicLink(link, target) {
```

emilyyyylime replied:
I don't really care for this; you can commut it if you want

YHerm replied:
commit suggestion doesn't refactor

YHerm commented on build.gradle:
```diff
@@ -148,33 +149,60 @@ tasks.register('checkAkitInstall', JavaExec) {
+
+        Files.createSymbolicLink(link, target)
+
+        def relative = { f -> rootDir.toPath().relativize(f) }
```
maybe move it to outside function, "getPathFromRootDir(link)"

emilyyyylime replied:
I personally disagree, but thinking about it; what do you think of just foregoing any printing there at all? Most tasks aren't so verbose

YHerm replied:
what do you mean? 
Also about the outside function, when its outside it can be used more easily in other tasks and more readable and informative

emilyyyylime replied:
I mean what if we don't print the whole 'Installed symbolic link ...'; it's not that necessary

YHerm commented on build.gradle:
```diff
@@ -148,33 +149,60 @@ tasks.register('checkAkitInstall', JavaExec) {
+}
+
+def gitConfigTasks = tasks.matching { it.group == gitConfig }
+tasks
+        .matching { it.group != gitConfig }
+        .configureEach { it.dependsOn gitConfigTasks }
```
i don't like this space, i would do it 1 line or 2

emilyyyylime replied:
it just doesn't really look good with the dots on different columns, but if it really bothers you two lines is certainly more readable than one

YHerm replied:
so 2 please

YHerm commented on build.gradle:
```diff
@@ -148,33 +149,60 @@ tasks.register('checkAkitInstall', JavaExec) {
+def gitConfig = 'Git Configuration'
+tasks.register('installLocalGitHookSymlinks') {
+    group = gitConfig
+
+    outputs.dir "$rootDir/.git/hooks/"
+
     doLast {
-        println("---Added Pre Commit Git Hook Script---")
+        def linkDir = file("$rootDir/.git/hooks/").toPath()
+        def gitHooks = fileTree("$rootDir/.scripts/git-hooks/")
+        gitHooks.forEach { File target ->
+            def link = linkDir.resolve(target.toPath().getFileName())
```
maybe save:
hooksDir = "$rootDir/.git/hooks/"
scriptsDir = "$rootDir/.scripts/git-hooks/"

and use the variables in the function

emilyyyylime commented:
I force-pushed this branch to implement minimal verification after every commit so we can finally properly merge this PR.
(So make sure to force pull if making local changes :heart:)

YHerm commented on .scripts/git-hooks/post-commit:
```diff
@@ -0,0 +1,18 @@
+BOLD=$(printf '\033[1m')
+RESET=$(printf '\033[0m')
+
+# Run spotlessCheck on committed files
```
no need

YHerm commented on build.gradle:
```diff
@@ -1,5 +1,8 @@
 import edu.wpi.first.gradlerio.GradleRIOPlugin
 import groovy.json.JsonSlurper
+import java.nio.file.Files
+import java.nio.file.NotLinkException
+import java.io.FileFilter
```
spotless should remove these

emilyyyylime replied:
I don't think it possibly could, it doesn't know which imports are actually used. But good catch

YHerm commented on build.gradle:
```diff
@@ -142,19 +145,55 @@ tasks.withType(JavaCompile).configureEach {
+
+// Setup local git hooks
+
+def ensureSymlink(link, target) {
```
```suggestion
def ensureSymbolicLink(link, target) {
```

emilyyyylime replied:
Again, Symlink is the much more common name for symbolic links. It's like calling tabs 'tabulations'.

YHerm commented on build.gradle:
```diff
@@ -142,19 +145,55 @@ tasks.withType(JavaCompile).configureEach {
-		unix(0775)
+
+// Setup local git hooks
+
```
remove line space

YHerm commented on build.gradle:
```diff
@@ -142,19 +145,55 @@ tasks.withType(JavaCompile).configureEach {
+
+		Files.createSymbolicLink(link, target)
+
+		def relative = { f -> rootDir.toPath().relativize(f) }
```
what relative is doing? (find more informative name)

YHerm commented on build.gradle:
```diff
@@ -142,19 +145,55 @@ tasks.withType(JavaCompile).configureEach {
+
+		Files.createSymbolicLink(link, target)
+
+		def relative = { f -> rootDir.toPath().relativize(f) }
```
what is f?

emilyyyylime replied:
come on this closure is literally used twice ever and both of those are on the next line the whole idea is to refactor out the repetition so the template string can be cleaner. If I give everything long names I might as well just write the entire `relativize` in the template string

YHerm replied:
this is a function that gets path and make it relative to root, i think the naming of the function and f should represent it

YHerm replied:
you can call the func rootRelative and f path

emilyyyylime replied:
I called it `shortname` in the new code if you look, because that's the reason it's used here

YHerm replied:
yes i saw, but the function still gets path and make it relative to root and names should say that

YHerm commented on build.gradle:
```diff
@@ -142,19 +145,55 @@ tasks.withType(JavaCompile).configureEach {
+}
+
+def gitConfig = 'Git Configuration'
+tasks.register('installLocalGitHookSymlinks') {
```
```suggestion
tasks.register('installLocalGitHookSymbolicLinks') {
```

YHerm commented on build.gradle:
```diff
@@ -142,19 +145,55 @@ tasks.withType(JavaCompile).configureEach {
 	}
+}
+
+def gitConfig = 'Git Configuration'
```
```suggestion
def gitConfigGroupName = 'Git Configuration'
```

emilyyyylime replied:
I'm pretty sure it's more idiomatic to do it as I did. From usage, `group = gitConfig` is of course much nicer than `group = gitConfigGroupName`

YHerm commented on build.gradle:
```diff
@@ -142,19 +145,55 @@ tasks.withType(JavaCompile).configureEach {
+		def hooksDest = file("$rootDir/.git/hooks/").toPath()
+
+		hooksSources.forEach { File target ->
+			def link = hooksDest.resolve(target.toPath().getFileName())
```
can't you do one target just get name?

YHerm commented on build.gradle:
```diff
@@ -142,19 +145,55 @@ tasks.withType(JavaCompile).configureEach {
+
+		hooksSources.forEach { File target ->
+			def link = hooksDest.resolve(target.toPath().getFileName())
+
```
no need this line space

YHerm commented on build.gradle:
```diff
@@ -142,19 +145,55 @@ tasks.withType(JavaCompile).configureEach {
+		def hooksSources = fileTree("$rootDir/.scripts/git-hooks/")
+		def hooksDest = file("$rootDir/.git/hooks/").toPath()
+
+		hooksSources.forEach { File target ->
```
```suggestion
		hooksSources.forEach { File targetHookScript ->
```

YHerm commented on build.gradle:
```diff
@@ -142,19 +145,55 @@ tasks.withType(JavaCompile).configureEach {
+def gitConfigTasks = tasks.matching { it.group == gitConfig }
+tasks.matching({ it.group != gitConfig }).configureEach({ it.dependsOn gitConfigTasks })
+
+if (System.getenv().containsKey('CI') || !file('.git').isDirectory()) {
```
when the if itself run, on build, on refresh gradle?

emilyyyylime replied:
I believe on gradle refresh

emilyyyylime commented:
Sorry I must've missed your review

emilyyyylime commented:
Rebased to resolve conflicts

emilyyyylime commented:
@bossyshrimp68 heard you're available for CRs again. I figured you'd probably miss this one because you've been a pending reviewer for 2 months now—if you have the time, I'd appreciate another look over it :)

bossyshrimp68 commented on build.gradle:
```diff
@@ -142,19 +143,56 @@ tasks.withType(JavaCompile).configureEach {
-	filePermissions {
-		unix(0775)
+// Get a path as relative to the root directory of the project (usually for printing)
+def rootRelative(path) {
```
```suggestion
def getRootRelativePath(path) {
```

hillelv commented on build.gradle:
```diff
@@ -142,19 +143,56 @@ tasks.withType(JavaCompile).configureEach {
 	doLast {
-		println("---Added Post Checkout Git Hook Script---")
+		def hooksSources = fileTree("$rootDir/.scripts/git-hooks/")
+		def hooksDest = file("$rootDir/.git/hooks/").toPath()
```
destroy or destiny?

hillelv replied:
But it's more fun this way 🤡

emilyyyylime commented:
Rebased onto master and updated `gradlew` one more time (spotless was broken with the previous version). Neither of these changes should be accepted by reviewers so this is ready to merge

Nitay4 reviewed (dismissed):
gg you are sigma

emilyyyylime commented:
@bossyshrimp68 you the last one please approve :pray:

YHerm commented on build.gradle:
```diff
@@ -142,19 +143,56 @@ tasks.withType(JavaCompile).configureEach {
-	into { new File(rootProject.rootDir, '.git/hooks') }
-	filePermissions {
-		unix(0775)
+// Get a path as relative to the root directory of the project (usually for printing)
```
consider deleting this comment

YHerm commented on build.gradle:
```diff
@@ -142,19 +143,56 @@ tasks.withType(JavaCompile).configureEach {
+	rootDir.toPath().relativize(path)
+}
+
+// Setup local git hooks
+def ensureSymlink(link, target) {
```
maybe change func name to represent the comment instead of adding comment

emilyyyylime replied:
the comment is for that section of code, not just the function. I'll make that more obvious

YHerm commented on build.gradle:
```diff
@@ -142,19 +143,56 @@ tasks.withType(JavaCompile).configureEach {
 	doLast {
-		println("---Added Post Checkout Git Hook Script---")
+		def hooksSources = fileTree("$rootDir/.scripts/git-hooks/")
+		def hooksDestinationBecauseHillelInsisted = file("$rootDir/.git/hooks/").toPath()
```
this name looks a bit weird

YHerm commented on gradlew:
```diff
@@ -84,7 +86,8 @@ done
 APP_BASE_NAME=${0##*/}
 # Discard cd standard output in case $CDPATH is set (https://github.com/gradle/gradle/issues/25036)
-APP_HOME=$( cd "${APP_HOME:-./}" > /dev/null && pwd -P ) || exit
+APP_HOME=$( cd -P "${APP_HOME:-./}" > /dev/null && printf '%s
+' "$PWD" ) || exit
```
why 2 lines?

emilyyyylime replied:
Again this is not our code this is just an update to gradle because spotless didn't work with the other version

YHerm reviewed (dismissed):
U need to make an easy way to cancel it in competition, can be in another pr

emilyyyylime commented:
> U need to make an easy way to cancel it in competition, can be in another pr

As a reminder this is just a visual warning and does not prevent your commits from going through

YHerm commented on build.gradle:
```diff
@@ -142,19 +143,57 @@ tasks.withType(JavaCompile).configureEach {
-	into { new File(rootProject.rootDir, '.git/hooks') }
-	filePermissions {
-		unix(0775)
+// Setup local git hooks
```
i dont think its needed

YHerm commented on build.gradle:
```diff
@@ -142,19 +143,57 @@ tasks.withType(JavaCompile).configureEach {
+}
+
+
+// Set VisualVM host IP
```
no reason for this comment

YHerm commented on build.gradle:
```diff
@@ -142,19 +143,57 @@ tasks.withType(JavaCompile).configureEach {
+
 	doLast {
-		println("---Added Post Checkout Git Hook Script---")
+		def hooksSources = fileTree("$rootDir/.scripts/git-hooks/")
+		def hooksDestination = file("$rootDir/.git/hooks/").toPath()
```
cant u use here the getRootRelativePath?

emilyyyylime replied:
That does the opposite thing

emilyyyylime commented:
@YHerm is this good now? I don't want to ask others for reviews if you're gonna reject it anyways

YHerm reviewed (approved):
You won 😔

YHerm commented on build.gradle:
```diff
@@ -142,19 +143,54 @@ tasks.withType(JavaCompile).configureEach {
+tasks.register('installLocalGitHookSymlinks') {
+	group = gitConfig
+
+	inputs.dir "$rootDir/.scripts/git-hooks/"
+	inputs.dir "$rootDir/.git/hooks/"
+	outputs.dir "$rootDir/.git/hooks/"
```
consider to save the two paths as variables instead of duplicating it 5 times (another 2 times below)

YHerm commented on build.gradle:
```diff
@@ -142,19 +143,54 @@ tasks.withType(JavaCompile).configureEach {
+tasks.register('installLocalGitHookSymlinks') {
+	group = gitConfig
+
+	inputs.dir "$rootDir/.scripts/git-hooks/"
+	inputs.dir "$rootDir/.git/hooks/"
+	outputs.dir "$rootDir/.git/hooks/"
```
consider to save the two paths as variables instead of duplicating it 5 times (another 2 times below)

emilyyyylime replied:
"consider" or are you _requesting changes_? I'm getting mixed messages

YHerm replied:
I am suggesting what I see as a better option, when people does not agree they can answer and tell why and by that we get in the end the best solution

YHerm commented on build.gradle:
```diff
@@ -142,19 +143,56 @@ tasks.withType(JavaCompile).configureEach {
+	rootDir.toPath().relativize(path)
+}
+
+def ensureSymlink(link, target) {
```
add types

YHerm commented on build.gradle:
```diff
@@ -142,19 +143,56 @@ tasks.withType(JavaCompile).configureEach {
-	into { new File(rootProject.rootDir, '.git/hooks') }
-	filePermissions {
-		unix(0775)
+def getRootRelativePath(path) {
```
add types

emilyyyylime commented:
lmao
